Implement labels with async-profiler's contextId

korniltsev opened the pull request:
The api is:
```
Pyroscope.LabelsWrapper.run(new LabelsSet("thread_name", Thread.currentThread().getName()), () -> {
    // ...
});
```


See https://github.com/jvm-profiling-tools/async-profiler/pull/576 for reference

petethepig commented on agent/src/main/java/io/pyroscope/api/Tags.java:
```diff
@@ -0,0 +1,264 @@
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+
+public class Tags {
```
Similar to my comment in the other PR (https://github.com/pyroscope-io/pyroscope/pull/1096/files#r872029411), we should call these labels for the sake of consistency with other profilers.

korniltsev replied:
Renamed to 'Labels' and 'LabelsSet'

korniltsev replied:
Please let me know if you like new names

petethepig commented on agent/src/main/java/io/pyroscope/api/Tags.java:
```diff
@@ -0,0 +1,264 @@
+            AsyncProfiler.getInstance().setContextId(prevContextId);
+        }
+    }
+
```
extra new lines?

korniltsev replied:
removed

petethepig commented on agent/src/main/java/io/pyroscope/api/Tags.java:
```diff
@@ -0,0 +1,264 @@
+        final Long prevContextId;
+
+        public Scope(Object... args) { //todo create constructor for 2, 4, 6 args
+            if (args.length %2  != 0) {
```
```suggestion
            if (args.length %2 != 0) {
```

korniltsev replied:
I reformatted code by hand

petethepig commented on agent/src/main/java/io/pyroscope/api/Tags.java:
```diff
@@ -0,0 +1,264 @@
+
+        public Scope(Object... args) { //todo create constructor for 2, 4, 6 args
+            if (args.length %2  != 0) {
+                throw new IllegalArgumentException();
```
is it common to provide an error message in this case? (I'm a Java noob)

korniltsev replied:
I added error message

korniltsev commented:
Hi @petethepig 
I've addressed your comments.
I've also added code for uploading, demo app, small unit test, so you may want to take a look again.
Also this is still WIP, I am still working on implementing more concurrent contexts.

petethepig reviewed (approved):
LGTM!
